### ISSUE TITLE
Can send message but can't receive

### ISSUE THREAD

**Issue by fmulyono** [googlesamples/android-BluetoothChat#7]
Hi.

Thanks for the sample, it's been very helpful to get me started!
I tested the sample on two devices:

1. Nexus 5 (Android 6.0.1)
2. ASUS K019 (Android 4.4.2)

I paired both devices, and ASUS K019 can send message perfectly to my Nexus 5 (and received perfectly too).
But, my Nexus 5 can't seem to send message to ASUS device. There's no error message on Nexus, but ASUS isn't showing up any messages sent from Nexus device.

Any idea?

Thanks!

**Comment by fmulyono** [comment]
I just tested using two ASUS K019 (Android 4.4.2) devices.
(These two devices are identical)

Device A can send message to Device B (and received perfectly), but Device B can't send message to Device A.

**Comment by xXWingadiumXx** [comment]
Hi fmulyono,
I am a newcomer to android so sorry if my my post is not good enough.

I have tried the sample on these devices:
      1. Xiaomi HM 1W (Android ver 4.2.2)
      2. Samsung S3 ( Android ver 4.3)

The Devices can connect to each other, however it is always a one way chat.

I have researched on my own and found this link:
https://android-er.blogspot.sg/2016/04/problem-of-bluetoothchat-example.html

It just shows that changing the build SDK could solve some of the problem, but the root cause is unknown.

If anyone has any information please feel free to share. Thanks!

**Comment by sirmordred** [comment]
Give it a try https://github.com/sirmordred/android-BluetoothChat/commit/c9f8491e9fc9628071bfb8aa707a29786ecf9a24

**Comment by maxmalysh** [comment]
@sirmordred this works!

**Comment by ChenQingZhen** [comment]
@sirmordred Good！

**Comment by BlancGuillaume** [comment]
Thank you @sirmordred !!!!

**Comment by joaobiriba** [comment]
![](https://media.giphy.com/media/l2SpMXtpZg1gRfvmo/giphy.gif)

**Comment by anpoz** [comment]
![8pg 8o s6 a9hwda j 1xn](https://cloud.githubusercontent.com/assets/2825342/18190414/6b9b39c6-70f5-11e6-988b-ffc817e3c91a.gif)

**Comment by 1mfk1** [comment]
Can someone please explain why this worked?

**Comment by sirmordred** [comment]
Current state must be defined BEFORE starting ConnectedThread. Failure to do so opens a ***race condition*** - the "while (mState == STATE_CONNECTED)" loop in run() may exit because mState is not yet current. Particularly nasty, since it doesn't ALWAYS happen.
also more complete solution is https://github.com/googlesamples/android-BluetoothChat/pull/9

**Comment by luigidifraia** [comment]
@sirmordred Please see my comment for issue #8 (duplicate). Is the subtitle of the action bar correctly set upon connection or do you get a "connected to null" message?

**Comment by sirmordred** [comment]
@luigidifraia  yeah, experienced the same issue after resolving race condition..and fixed it in another commit
see https://github.com/googlesamples/android-BluetoothChat/pull/9
all the patches in PR should be applied in order to get correct action bar changes and logs in logcat

**Comment by luigidifraia** [comment]
@sirmordred Cool. I see the PR is now closed, which explains it. It's a shame it's not yet merged though. I will keep working on my own fork for now.

**Comment by luigidifraia** [comment]
@sirmordred I had a throughout look at PR #9 and here's some feedback I'd like to share:
- Thank you for the excellent work on fixing those 2 race conditions and the rest of the rationalisations :)
- BluetoothChatService.java's line 88 has no real effect. Was it a debug instruction to set a breakpoint or something along those lines?
- I seem to understand the setState() method itself might have been initially introduced in order to log state changes at the same time as they were requested: if it were me, I would not have included the functionality to also message back the UI Activity about the change, but that was probably done for convenience and in order to avoid code repetition - i.e. one cannot forget to update the UI Activity after a state change and the code to do so is in a single location. You opted for local changes to mState instead of using a setState() method but are now logging state changes in updateUserInterfaceTitle() and messaging back the UI Activity: again this is something that mixes a logging task about state changes and an UI update task, for convenience, although it feels like the method name is misleading.
I would go for something along these lines, with a method name that explicitly states both purposes and also addresses the fact the device name needs to be shared prior to advertising the state changing to STATE_CONNECTED (as we commented in Issue #8):
```java
    /**
     * Set the current state of the chat connection, log it, and share it
     * with the UI Activity
     *
     * @param state An integer defining the current connection state
     */
    private synchronized void setStateAndUpdateUI(int state) {
        Log.d(TAG, "setState() " + mState + " -> " + state);
        mState = state;

        if (state == STATE_CONNECTED) {
            // Send the name of the connected device back to the UI Activity
            Message msg = mHandler.obtainMessage(Constants.MESSAGE_DEVICE_NAME);
            Bundle bundle = new Bundle();
            bundle.putString(Constants.DEVICE_NAME, device.getName());
            msg.setData(bundle);
            mHandler.sendMessage(msg);
        }

        // Give the new state to the Handler so the UI Activity can update
        mHandler.obtainMessage(Constants.MESSAGE_STATE_CHANGE, state, -1).sendToTarget();
    }
```
Invocation of the above method would occur everywhere you set mState directly. All calls to updateUserInterfaceTitle() would be removed.

Any thoughts?
